### PR TITLE
[core] Convert token strings to expressions

### LINKED ISSUE #11659: Serialize tokens in expressions as get expressions
A serialized expression should never contain token syntax. As of #11509 and #11452, if style JSON contains a `text-field` property set to a legacy function, the runtime styling getter for that property returns an expression containing strings with tokens in them. So simply getting a `text-field` then setting `text-field` to the same value will turn tokens into literal `{token}`s on screen.

For #11651, I’m working around this issue by implementing an “upgrade” step in Objective-C that replaces tokens with `mgl_join:` and key path expressions, but an mbgl implementation would be somewhat more performant and benefit more platforms.

/cc @anandthakker

### PR DESCRIPTION
**Pull request by jfirebaugh**
Fixes #11659.

This may need some additional coordination with SDK code. Specifically, the constructor `DataDrivenPropertyValue(T)` (where `T` is `std::string`) no longer accepts token strings for `icon-image` or `text-field`. Such strings should be converted to expressions, and then the `DataDrivenPropertyValue(PropertyExpression<T>)` constructor should be used. It's difficult to tell if the former constructor is in use by SDK code (I can't simply delete it and see what fails to compile, because there are lots of uses that aren't affected).

### CODE REVIEW

**Comment by 1ec5** [comment]
Per https://github.com/mapbox/mapbox-gl-native/issues/11659#issuecomment-405686351, it’s likely that the iOS and macOS map SDKs still support setting properties to constant string expressions containing `{token}` syntax. However, we no longer document that syntax, and removing support for it could be chalked up to a fix for a roundtripping bug:

> Fixed an issue where some constant value expressions were treated as key path expressions. It is now possible to set the `MGLSymbolStyleLayer.text` property to a string that literally contains a word in curly braces.

**Comment by 1ec5** [comment]
> we no longer document the token syntax

[I stand corrected](https://www.mapbox.com/ios-sdk/api/4.1.1/Classes/MGLSymbolStyleLayer.html#/c:objc(cs)MGLSymbolStyleLayer(py)iconImageName). Nonetheless, we could implement the reverse of the transformation described in https://github.com/mapbox/mapbox-gl-native/issues/11659#issuecomment-405686351 at the SDK level to unblock this work: scan the string for curly braces and convert to key path expressions. The downside is that the SDK has no context about the feature’s fields, so #11953 would effectively regress.

**Comment by jfirebaugh** [comment]
> Nonetheless, we could implement the reverse of the transformation described in #11659 (comment) at the SDK level to unblock this work: scan the string for curly braces and convert to key path expressions.

I've done so in the latest commit; could you please take a look?

> The downside is that the SDK has no context about the feature’s fields, so #11953 would effectively regress.

#11953 applied only to URLs; tokens for non-existent feature properties have always been [replaced with an empty string](https://github.com/mapbox/mapbox-gl-native/blob/bffc60fefda049c8cde0843988c2b73ab87b6103/src/mbgl/layout/symbol_layout.cpp#L111-L112).

**Comment by jfirebaugh** [comment]
@tobrun How about Android? Does the SDK need to continue to support e.g. `symbolLayer.setProperties(iconImage("{token}"))`? In contrast to iOS, this [doesn't seem to be documented](https://www.mapbox.com/android-docs/api/map-sdk/6.2.1/com/mapbox/mapboxsdk/style/layers/PropertyFactory.html#iconImage-java.lang.String-) for Android.

File: platform/darwin/src/MGLStyleLayer.mm.ejs [hunk 1]
```diff
@@ -121,6 +121,16 @@ namespace mbgl {
 - (void)set<%- camelize(property.name) %>:(NSExpression *)<%- objCName(property) %> {
     MGLAssertStyleLayerIsValid();
 
+<% if (property.tokens) { -%>
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: This part of the template is only for layout properties. If paint properties like `background-pattern` can also accept tokens, then this change needs to be repeated further down in the template.

Comment (jfirebaugh) [reply platform/darwin/src/MGLStyleLayer.mm.ejs]: `text-field` and `icon-image`, both layout properties, are the only ones that accept tokens, and we don't plan to add any token-accepting properties in the future.

**Comment by tobrun** [comment]
> @tobrun How about Android? Does the SDK need to continue to support e.g. symbolLayer.setProperties(iconImage("{token}"))? In contrast to iOS, this doesn't seem to be documented for Android.

Long term, I feel not as afaik that use-case is completely covered with the `get` expression. For now, since next release isn't planned to be a major release, it seems that we should as it can potentially break existing client code. 

Our documentation is indeed a little sparse on the token syntax, we only show case usage through examples. I have been trying to up our documentation game with the introduction of expressions.

**Comment by jfirebaugh** [comment]
Looks like Android will continue to support `setProperties(iconImage("{token}"))` without any additional changes. I'll add a couple of tests to confirm.

**Comment by tobrun** [comment]
@jfirebaugh did a quick manual test on this PR and still works 👍

File: test/style/conversion/function.test.cpp [hunk 2]
```diff
@@ -75,3 +76,20 @@ TEST(StyleConversion, CompositeFunctionExpression) {
...
+    ASSERT_TRUE(hasTokens("token {token}"));
+    ASSERT_TRUE(hasTokens("{token} {token}"));
+
+    using namespace mbgl::style::expression::dsl;
+    ASSERT_EQ(*convertTokenStringToExpression("{token}"), *toString(get(literal("token"))));
+    ASSERT_EQ(*convertTokenStringToExpression("token {token}"), *concat(vec(literal("token "), toString(get(literal("token"))))));
+    ASSERT_EQ(*convertTokenStringToExpression("{token} token"), *concat(vec(toString(get(literal("token"))), literal(" token"))));
+    ASSERT_EQ(*convertTokenStringToExpression("{token} {token}"), *concat(vec(toString(get(literal("token"))), literal(" "), toString(get(literal("token"))))));
+    ASSERT_EQ(*convertTokenStringToExpression("{token} {token"), *concat(vec(toString(get(literal("token"))), literal(" "), literal("{token"))));
+    ASSERT_EQ(*convertTokenStringToExpression("{token {token}"), *concat(vec(literal("{token "), toString(get(literal("token"))))));
```
Comment (anandthakker) [inline test/style/conversion/function.test.cpp]: Maybe add `"{token}{token}"` as well

**Pull request merged**